### PR TITLE
Fixes the segmentation Fault in Capped distance

### LINKED ISSUE #2164: Strange segfault in capped_distance
I'm getting a strange intermittent segfault when using `capped_distance`.  This is basically searching for all pairs within 5.0 of different AtomGroups.  Running the code below will segfault for different pairs each time as well....

```python
import MDAnalysis as mda
import itertools
from MDAnalysis.lib.distances import capped_distance

u = mda.Universe('single.dcd')

nfrags = 200
apf = len(u.atoms) // nfrags
frags = [u.atoms[apf * i:apf * (i+1)] for i in range(nfrags)]


for i, j in itertools.combinations(range(nfrags), 2):
    frag_i, frag_j = frags[i], frags[j]

    print(frag_i.indices.min(), frag_j.indices.min())
    capped_distance(frag_i.positions, frag_j.positions, 5.0)
```

### PR DESCRIPTION
**Pull request by ayushsuhane**
Fixes #2164 

Changes made in this Pull Request:
 - Modified the NSGrid to handle the case when probe exactly touches the boundary of the box. 

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/lib/nsgrid.pyx [hunk 1]
```diff
@@ -901,6 +901,8 @@ cdef class FastNS(object):
                                     if probe[m] > self.box.c_pbcbox.box[m][m]:
                                         check = False
                                         break
+                                    if probe[m] == self.box.c_pbcbox.box[m][m]:
```
Comment (orbeckst) [inline package/MDAnalysis/lib/nsgrid.pyx]: I am always wary of testing equality for floats. Is this a case when it is justified to do so?

Or should this also do a comparison within a certain epsilon?

File: package/MDAnalysis/lib/nsgrid.pyx [hunk 2]
```diff
@@ -901,6 +901,8 @@ cdef class FastNS(object):
...
+                                        probe[m] -= 0.01*self.grid.cellsize[m]
```
Comment (orbeckst) [inline package/MDAnalysis/lib/nsgrid.pyx]: What's the justification for 0.01? Does the exact value matter? Could it be made smaller?

File: package/MDAnalysis/lib/nsgrid.pyx [hunk 3]
```diff
@@ -979,9 +981,11 @@ cdef class FastNS(object):
                                     if probe[m] < 0:
                                         check = False
                                         break
-                                    elif probe[m] >= self.box.c_pbcbox.box[m][m]:
+                                    elif probe[m] > self.box.c_pbcbox.box[m][m]:
                                         check = False
                                         break
+                                    if probe[m] == self.box.c_pbcbox.box[m][m]:
```
Comment (orbeckst) [inline package/MDAnalysis/lib/nsgrid.pyx]: see comments above

**Comment by richardjgowers** [comment]
@ayushsuhane thanks for the fix, but I think this is fixing the issue in the wrong locatio - `coord2cellid` shouldn't return invalid cellids ever. See #2175

File: package/MDAnalysis/lib/nsgrid.pyx (hunk 1 above)
Comment (ayushsuhane) [reply package/MDAnalysis/lib/nsgrid.pyx]: Yes @orbeckst, you are right. It is not a good idea to compare the equality of floats. While debugging, I found the error here which seemed to get resolved by this correction. But I agree, it wasn't robust enough.

Comment (ayushsuhane) [reply package/MDAnalysis/lib/nsgrid.pyx]: As far as the algorithm is concerned, this specific patch tries to find the neighboring cell by moving in all 3 directions (27 combinations using probe). The problem here was that probe was moving out of the box dimensions (not quite, very close to the edge of the box).

**Comment by ayushsuhane** [comment]
Thanks @richardjgowers #2175 does the correct treatment to not allow invalid cell ids. However, I figured the problem was specifically due to the artificial box creation. Earlier, the box was created with 1.2 times the  (minimum - maximum) dimensions in either directions. But it implicitly assumed that the cutoff would be 10% of the box size. I think it was the reason the error was not reproducible. Now I changed the box creation in a way that the new box's max dimension would be atleast "cutoff" distance away from  the nearest particle. Hope it gives away the right picture, and also the right solution. (Time shall tell :P)

**Comment by richardjgowers** [comment]
Ok finally got round to checking, this works a charm, thanks @ayushsuhane (and @orbeckst for the poke)

**Pull request merged**